File: counting_sort/counting_sort.go
package main

import "fmt"

// CountingSort sorts the input array in ascending order
// Time complexity is O(n)
func CountingSort(data []int) {
	temp := make([]int, len(data))
	max := 0
	for _, elem := range data {
		if elem > max {
			max = elem
		}
	}
	count := make([]int, max+1)
	for _, item := range data {
		count[item]++
	}
	for i := 1; i <= max; i++ {
		count[i] += count[i-1]
	}
	for i := 0; i < len(data); i++ {
		temp[count[data[i]]-1] = data[i]
		count[data[i]]--
	}
	for i := 0; i < len(data); i++ {
		data[i] = temp[i]
	}
}

func main() {
	data := []int{1, 202, 2, 675, 901, 116, 312, 1, 2}
	CountingSort(data)
	fmt.Println(data)
}

File: counting_sort/CountingSort.cs
using System;

public class CountingSort
{

    /**
    * Sorts the array using counting sort
    * Time Complexity: O(n + k) where n is the number of elements in the array and k is the range of input
    * Space Complexity: O(n + k)
    */
    public static int[] DoCountingSort(int[] arr)
    {
        // finding the maximum value in the array
        int maxValue = Int32.MinValue;
        foreach (int num in arr)
            maxValue = Math.Max(maxValue, num);

        // counting the frequency of each element in the array
        int[] frequency = new int[maxValue + 1];
        foreach (int num in arr)
            frequency[num]++;

        // modifying each element in the frequency array to the total of all the elements before it
        for (int i = 1; i < frequency.Length; i++)
            frequency[i] += frequency[i - 1];

        // constructing the sorted array based on the freqency of each element in ascending order
        int[] sortedArr = new int[arr.Length];
        for (int i = 0; i < arr.Length; i++) {
            sortedArr[frequency[arr[i]] - 1] = arr[i];
            frequency[arr[i]]--;
        }
        return sortedArr;
    }

    public static void Main()
    {
        int[] arr = new int[] {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
        int[] sortedArr = DoCountingSort(arr);
        foreach (int num in sortedArr)
            Console.WriteLine(num);
    }
}
